### ISSUE TITLE
unexpected fault address on raspberry pi4 and openwrt

### ISSUE THREAD

**Issue by sashasimkin** [prometheus/prometheus#8589]
**What did you do?**
Installed prometheus from openwrt repositories.
Running /etc/init.d/prometheus didn't yield expected results and I couldn't find any logs, so I inspected the script and started it manually as:
```
prometheus --config.file=/etc/prometheus.yml --storage.tsdb.path=/data --web.listen-address=127.0.0.1:9090
```

**What did you expect to see?**
Running prometheus instance, metrics in grafana cloud.

**What did you see instead? Under which circumstances?**
Prometheus is crashing with segfault after a few seconds of running, I suppose while trying to remote_write the data.

**Environment**

* System information:
```
Linux 5.4.102 aarch64
```

* Prometheus version:
```
prometheus, version v2.22.0 (branch: release, revision: 2.22.0)
  build user:       openwrt
  build date:       1615203910
  go version:       go1.16
  platform:         linux/arm64
```

* Prometheus configuration file:
```
# cat /etc/prometheus.yml
global:
  scrape_interval:     15s # Set the scrape interval to every 15 seconds. Default is every 1 minute.
  evaluation_interval: 15s # Evaluate rules every 15 seconds. The default is every 1 minute.
  # scrape_timeout is set to the global default (10s).

# Alertmanager configuration
alerting:
  alertmanagers:
  - static_configs:
    - targets:
      # - alertmanager:9093

# Load rules once and periodically evaluate them according to the global 'evaluation_interval'.
rule_files:
  # - "first_rules.yml"
  # - "second_rules.yml"

# A scrape configuration containing exactly one endpoint to scrape:
# Here it's Prometheus itself.
scrape_configs:
  # The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
  - job_name: 'prometheus'

    # metrics_path defaults to '/metrics'
    # scheme defaults to 'http'.

    static_configs:
    - targets: ['localhost:9090']

  - job_name: 'openwrt'
    static_configs:
    - targets: ['localhost:9100']

remote_write:
- url: https://prometheus-us-central1.grafana.net/api/prom/push
  basic_auth:
    username: ~snip~
    password: ~snip~
```

* Logs:
```
# prometheus --config.file=/etc/prometheus.yml --storage.tsdb.path=/data --web.listen-address=127.0.0.1:9090
level=info ts=2021-03-11T22:27:05.125Z caller=main.go:315 msg="No time or size retention was set so using the default time retention" duration=15d
level=info ts=2021-03-11T22:27:05.125Z caller=main.go:353 msg="Starting Prometheus" version="(version=v2.22.0, branch=release, revision=2.22.0)"
level=info ts=2021-03-11T22:27:05.126Z caller=main.go:358 build_context="(go=go1.16, user=openwrt, date=1615203910)"
level=info ts=2021-03-11T22:27:05.126Z caller=main.go:359 host_details="(Linux 5.4.102 #0 SMP Sat Mar 6 19:23:48 2021 aarch64 OpenWrt (none))"
level=info ts=2021-03-11T22:27:05.126Z caller=main.go:360 fd_limits="(soft=1024, hard=4096)"
level=info ts=2021-03-11T22:27:05.126Z caller=main.go:361 vm_limits="(soft=unlimited, hard=unlimited)"
level=info ts=2021-03-11T22:27:05.133Z caller=main.go:712 msg="Starting TSDB ..."
level=info ts=2021-03-11T22:27:05.133Z caller=web.go:516 component=web msg="Start listening for connections" address=127.0.0.1:9090
level=info ts=2021-03-11T22:27:05.163Z caller=head.go:642 component=tsdb msg="Replaying on-disk memory mappable chunks if any"
level=info ts=2021-03-11T22:27:05.164Z caller=head.go:656 component=tsdb msg="On-disk memory mappable chunks replay completed" duration=13.778µs
level=info ts=2021-03-11T22:27:05.164Z caller=head.go:662 component=tsdb msg="Replaying WAL, this may take a while"
level=info ts=2021-03-11T22:27:05.165Z caller=head.go:714 component=tsdb msg="WAL segment loaded" segment=0 maxSegment=0
level=info ts=2021-03-11T22:27:05.165Z caller=head.go:719 component=tsdb msg="WAL replay completed" checkpoint_replay_duration=134.108µs wal_replay_duration=1.096238ms total_replay_duration=1.411956ms
level=info ts=2021-03-11T22:27:05.174Z caller=main.go:732 fs_type=EXT4_SUPER_MAGIC
level=info ts=2021-03-11T22:27:05.175Z caller=main.go:735 msg="TSDB started"
level=info ts=2021-03-11T22:27:05.175Z caller=main.go:861 msg="Loading configuration file" filename=/etc/prometheus.yml
ts=2021-03-11T22:27:05.178Z caller=dedupe.go:112 component=remote level=info remote_name=ddc323 url=https://prometheus-us-central1.grafana.net/api/prom/push msg="Starting WAL watcher" queue=ddc323
ts=2021-03-11T22:27:05.179Z caller=dedupe.go:112 component=remote level=info remote_name=ddc323 url=https://prometheus-us-central1.grafana.net/api/prom/push msg="Replaying WAL" queue=ddc323
level=info ts=2021-03-11T22:27:05.182Z caller=main.go:892 msg="Completed loading of configuration file" filename=/etc/prometheus.yml totalDuration=7.603544ms remote_storage=1.829077ms web_handler=2.259µs query_engine=3.092µs scrape=3.308586ms scrape_sd=381.048µs notify=274.235µs notify_sd=110.998µs rules=6.314µs
level=info ts=2021-03-11T22:27:05.183Z caller=main.go:684 msg="Server is ready to receive web requests."
ts=2021-03-11T22:27:16.930Z caller=dedupe.go:112 component=remote level=info remote_name=ddc323 url=https://prometheus-us-central1.grafana.net/api/prom/push msg="Done replaying WAL" duration=11.752045728s
unexpected fault address 0xa626f6a0336fd
fatal error: fault
[signal SIGSEGV: segmentation violation code=0x1 addr=0xa626f6a0336fd pc=0xa76060]

goroutine 250 [running]:
runtime.throw(0x20cb426, 0x5)
        runtime/panic.go:1117 +0x54 fp=0x4000e15560 sp=0x4000e15530 pc=0x436664
runtime.sigpanic()
        runtime/signal_unix.go:741 +0x230 fp=0x4000e155a0 sp=0x4000e15560 pc=0x44e080
github.com/golang/snappy.encodeBlock(0x40010c8003, 0xcb60, 0xcb60, 0x40010ba000, 0xae3a, 0xe000, 0xa)
        github.com/golang/snappy@v0.0.2/encode_arm64.s:666 +0x360 fp=0x4000e1d640 sp=0x4000e155b0 pc=0xa76060
github.com/golang/snappy.Encode(0x40010c8000, 0xcb63, 0xcb63, 0x0, 0x0, 0x0, 0xae3a, 0x4000f4417b, 0x4000b8f7a8)
        github.com/golang/snappy@v0.0.2/encode.go:39 +0x17c fp=0x4000e1d690 sp=0x4000e1d640 pc=0xa7559c
github.com/prometheus/prometheus/tsdb/wal.(*WAL).log(0x4000696360, 0x40010ba000, 0xae3a, 0xe000, 0x1, 0x17c, 0x4000624370)
        github.com/prometheus/prometheus/tsdb/wal/wal.go:622 +0x368 fp=0x4000e1d730 sp=0x4000e1d690 pc=0xa7cc98
github.com/prometheus/prometheus/tsdb/wal.(*WAL).Log(0x4000696360, 0x4000e1d838, 0x1, 0x1, 0x0, 0x0)
        github.com/prometheus/prometheus/tsdb/wal/wal.go:584 +0xc8 fp=0x4000e1d7c0 sp=0x4000e1d730 pc=0xa7c878
github.com/prometheus/prometheus/tsdb.(*headAppender).log(0x400054aa00, 0x0, 0x0)
        github.com/prometheus/prometheus/tsdb/head.go:1184 +0x26c fp=0x4000e1d870 sp=0x4000e1d7c0 pc=0xa9f10c
github.com/prometheus/prometheus/tsdb.(*headAppender).Commit(0x400054aa00, 0x0, 0x0)
        github.com/prometheus/prometheus/tsdb/head.go:1204 +0x74 fp=0x4000e1d9b0 sp=0x4000e1d870 pc=0xa9f264
github.com/prometheus/prometheus/tsdb.dbAppender.Commit(0x2488600, 0x400054aa00, 0x40007bc000, 0x3, 0x256065fde24b4dba)
        github.com/prometheus/prometheus/tsdb/db.go:758 +0x30 fp=0x4000e1d9f0 sp=0x4000e1d9b0 pc=0xa90760
github.com/prometheus/prometheus/tsdb.(*dbAppender).Commit(0x4000e88ab0, 0x400004f700, 0x400004f700)
        <autogenerated>:1 +0x48 fp=0x4000e1da30 sp=0x4000e1d9f0 pc=0xabb138
github.com/prometheus/prometheus/storage.(*fanoutAppender).Commit(0x400004f980, 0x2487fe0, 0x4000e88ac8)
        github.com/prometheus/prometheus/storage/fanout.go:176 +0x34 fp=0x4000e1dae0 sp=0x4000e1da30 pc=0xa324f4
github.com/prometheus/prometheus/scrape.(*timeLimitAppender).Commit(0x4000e88ac8, 0x178, 0x178)
        <autogenerated>:1 +0x3c fp=0x4000e1db10 sp=0x4000e1dae0 pc=0xb4972c
github.com/prometheus/prometheus/scrape.(*scrapeLoop).scrapeAndReport.func1(0x4000e1dce0, 0x4000b8fcf0, 0x4000cf0000)
        github.com/prometheus/prometheus/scrape/scrape.go:1074 +0x40 fp=0x4000e1db90 sp=0x4000e1db10 pc=0xb47d60
github.com/prometheus/prometheus/scrape.(*scrapeLoop).scrapeAndReport(0x4000cf0000, 0x37e11d600, 0x2540be400, 0x0, 0x0, 0x0, 0xc00ac431f862e235, 0x382867025, 0x32b6ce0, 0x0, ...)
        github.com/prometheus/prometheus/scrape/scrape.go:1141 +0x7a4 fp=0x4000e1de00 sp=0x4000e1db90 pc=0xb41144
github.com/prometheus/prometheus/scrape.(*scrapeLoop).run(0x4000cf0000, 0x37e11d600, 0x2540be400, 0x0)
        github.com/prometheus/prometheus/scrape/scrape.go:1027 +0x268 fp=0x4000e1dfb0 sp=0x4000e1de00 pc=0xb40688
runtime.goexit()
        runtime/asm_arm64.s:1130 +0x4 fp=0x4000e1dfb0 sp=0x4000e1dfb0 pc=0x46d974
created by github.com/prometheus/prometheus/scrape.(*scrapePool).sync
        github.com/prometheus/prometheus/scrape/scrape.go:497 +0x724

goroutine 1 [chan receive]:
github.com/oklog/run.(*Group).Run(0x4000f1db50, 0x40004c6460, 0x8)
        github.com/oklog/run@v1.1.0/group.go:43 +0xd4
main.main()
        github.com/prometheus/prometheus/cmd/prometheus/main.go:797 +0x6088
```

**Comment by volgar1x** [comment]
Same problem on `FreeBSD 14.0-CURRENT`

**Comment by roidelapluie** [comment]
Could you try with a prometheus built with go 1.15?

**Comment by volgar1x** [comment]
> Could you try with a prometheus built with go 1.15?

Cross-building from linux/amd64 into freebsd/arm64 with Go 1.15 fixed the problem. Many thanks!

**Comment by roidelapluie** [comment]
Does it happen all the time? Could you try go.16.2 to see if that has been fixed?

**Comment by sashasimkin** [comment]
Happened all the time for me w/ go 1.16. I've downloaded `prometheus-2.22.0.linux-arm64.tar.gz` from releases page, and it works well now!
```
root@OpenWrt:/tmp# ./prometheus --version
prometheus, version 2.22.0 (branch: HEAD, revision: 0a7fdd3b76960808c3a91d92267c3d815c1bc354)
  build user:       root@6321101b2c50
  build date:       20201015-13:29:53
  go version:       go1.15.3
  platform:         linux/arm64
```
Sorry, I'm not sure how to compile it with go 1.16.2 :(

**Comment by roidelapluie** [comment]
You could get a fresh binary for our main branch, in our ci, e.g.

https://app.circleci.com/pipelines/github/prometheus/prometheus/11251/workflows/0203e33d-4d95-42dc-a8ba-e08b7f07e8a1/jobs/49308/artifacts

(Remember that this is a test build)

**Comment by roidelapluie** [comment]
I suspect that https://github.com/prometheus/prometheus/pull/8593 did fix this issue. cc @prymitive

**Comment by roidelapluie** [comment]
I will close this as we have released v2.25.1 with a fix.

**Comment by sashasimkin** [comment]
2.25.2 confirmed working well on OpenWrt snapshot, build with go 1.16, ref: https://github.com/openwrt/packages/issues/15236#issuecomment-807802776

thank you!
